### PR TITLE
FEAT [`Generation`]: Introduce a centralized API to switch between cache implementations

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

I would like to introduce a new API before the release to centralize switching between cache implementations ! 

Right now to load SInkCache one needs to do:

```python
from transformers import AutoTokenizer, AutoModelForCausalLM, SinkCache

tokenizer = AutoTokenizer.from_pretrained("TheBloke/LLaMa-7B-GPTQ")
model = AutoModelForCausalLM.from_pretrained("TheBloke/LLaMa-7B-GPTQ", device_map="auto")

cache = SinkCache(window_length=508, num_sink_tokens=4)

inputs = tokenizer(["Vaswani et al. (2017) introduced the Transformers"], return_tensors="pt").to(model.device)
gen_out = model.generate(**inputs, do_sample=False, max_new_tokens=300, past_key_values=cache)
decoded = tokenizer.batch_decode(gen_out, skip_special_tokens=True)
```

For static cache: 

```python
from transformers import AutoTokenizer, AutoModelForCausalLM, SinkCache

tokenizer = AutoTokenizer.from_pretrained("TheBloke/LLaMa-7B-GPTQ")
model = AutoModelForCausalLM.from_pretrained("TheBloke/LLaMa-7B-GPTQ", device_map="auto")

model.generation_config.cache_implementation = "static"

inputs = tokenizer(["Vaswani et al. (2017) introduced the Transformers"], return_tensors="pt").to(model.device)
gen_out = model.generate(**inputs, do_sample=False, max_new_tokens=300)
decoded = tokenizer.batch_decode(gen_out, skip_special_tokens=True)
```

With this PR: 

```diff
from transformers import AutoTokenizer, AutoModelForCausalLM, SinkCache

tokenizer = AutoTokenizer.from_pretrained("TheBloke/LLaMa-7B-GPTQ")
model = AutoModelForCausalLM.from_pretrained("TheBloke/LLaMa-7B-GPTQ", device_map="auto")

- cache = SinkCache(window_length=508, num_sink_tokens=4)
+ model.set_cache_implementation("sink", sink_window_length=508, num_sink_tokens=4)

inputs = tokenizer(["Vaswani et al. (2017) introduced the Transformers"], return_tensors="pt").to(model.device)
- gen_out = model.generate(**inputs, do_sample=False, max_new_tokens=300, past_key_value=cache)
+ gen_out = model.generate(**inputs, do_sample=False, max_new_tokens=300)
decoded = tokenizer.batch_decode(gen_out, skip_special_tokens=True)
```

```diff
from transformers import AutoTokenizer, AutoModelForCausalLM, SinkCache

tokenizer = AutoTokenizer.from_pretrained("TheBloke/LLaMa-7B-GPTQ")
model = AutoModelForCausalLM.from_pretrained("TheBloke/LLaMa-7B-GPTQ", device_map="auto")

- model.generation_config.cache_implementation = "static"
+ model.set_cache_implementation("static")

inputs = tokenizer(["Vaswani et al. (2017) introduced the Transformers"], return_tensors="pt").to(model.device)
gen_out = model.generate(**inputs, do_sample=False, max_new_tokens=300, past_key_values=cache)
decoded = tokenizer.batch_decode(gen_out, skip_special_tokens=True)
```

What do you think @gante @tomaarsen @ArthurZucker @amyeroberts  ?

If you are happy with the design and idea I can move forward with adding tests and docs !

### CODE REVIEW

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -351,6 +359,43 @@ def prepare_inputs_for_generation(self, *args, **kwargs):
             "A model class needs to define a `prepare_inputs_for_generation` method in order to use `.generate()`."
         )
 
+    def switch_cache_implementation(self, cache_implementation: Union[CacheImplementation, str], **kwargs):
```
Comment (tomaarsen) [inline src/transformers/generation/utils.py:362]: In practice, we are obviously always switching from one cache implementation to another, but for the users it's likely more intuitive to simply `set` a new cache implementation, so `set_cache_implementation` might be a better method.

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -351,6 +359,43 @@ def prepare_inputs_for_generation(self, *args, **kwargs):
...
+        """
+        if cache_implementation.upper() not in CacheImplementation.__members__:
+            raise ValueError(
+                f"Unrecognized cache implementation - you passed {cache_implementation}. Supported cache implementations are"
+                f" {CacheImplementation.__members__}"
+            )
+
+        if not self._supports_cache_class:
+            raise ValueError(
+                "This model do not currently support switching between cache implementations. Please raise an error on GitHub for adding"
```
Comment (tomaarsen) [inline src/transformers/generation/utils.py:383]:
```suggestion
                "This model do not currently support switching between cache implementations. Please raise an issue on GitHub for adding"
```
or

Comment (tomaarsen) [reply src/transformers/generation/utils.py:383]:
```suggestion
                "This model do not currently support switching between cache implementations. Please raise a feature request on GitHub for adding"
```

File: src/transformers/generation/utils.py [hunk 3]
```diff
@@ -351,6 +359,43 @@ def prepare_inputs_for_generation(self, *args, **kwargs):
...
+                " the request to support this model: https://github.com/huggingface/transformers"
+            )
+
+        if cache_implementation == CacheImplementation.SINK:
+            if "window_length" not in kwargs and "num_sink_tokens" not in kwargs:
+                raise ValueError(
+                    "You asked to switch to Sink cache implementation but you did not passed `window_length` and `num_sink_tokens` to "
```
Comment (tomaarsen) [inline src/transformers/generation/utils.py:390]:
```suggestion
                    "You asked to switch to the Sink cache implementation, but you did not pass `window_length` and `num_sink_tokens` to "
```

**Review by tomaarsen — COMMENTED** [review]
Looks very clean at a glance! I have some small nits regarding phrasing & the method name.
Thanks for tackling this 🤗

File: src/transformers/generation/utils.py (hunk 1 above)
Comment (younesbelkada) [reply src/transformers/generation/utils.py:362]: I agree! thanks for ponting that out !

File: src/transformers/generation/utils.py [hunk 4]
```diff
@@ -351,6 +359,43 @@ def prepare_inputs_for_generation(self, *args, **kwargs):
             "A model class needs to define a `prepare_inputs_for_generation` method in order to use `.generate()`."
         )
 
+    def set_cache_implementation(self, cache_implementation: Union[CacheImplementation, str], **kwargs):
+        """
+        Simple API to switch cache implementation in a model. Users could also do
```
Comment (tomaarsen) [inline src/transformers/generation/utils.py:364]:
```suggestion
        Simple API to set cache implementation in a model. Users could also do
```

File: src/transformers/generation/utils.py [hunk 5]
```diff
@@ -351,6 +359,43 @@ def prepare_inputs_for_generation(self, *args, **kwargs):
...
+        `model.config.generation_config.cache_implementation = xxx` but they will most likely unexpected
+        behaviour
+
+        Args:
+            cache_implementation (`Union[CacheImplementation, str]`):
+                The target cache implementation
```
Comment (tomaarsen) [inline src/transformers/generation/utils.py:370]: Maybe include a list of the options? It might not be clear otherwise.

File: src/transformers/generation/utils.py [hunk 6]
```diff
@@ -351,6 +359,43 @@ def prepare_inputs_for_generation(self, *args, **kwargs):
...
+            kwargs (`dict`, *optional*):
+                Optional key word arguments to be passed. E.g. for SinkCache, it is required to
```
Comment (tomaarsen) [inline src/transformers/generation/utils.py:372]:
```suggestion
                Optional key word arguments to be passed. E.g. for "sink", it is required to
```
This might be clearer as we only deal with e.g. "static", "sink" and "dynamic" in this function

File: src/transformers/generation/utils.py [hunk 7]
```diff
@@ -351,6 +359,43 @@ def prepare_inputs_for_generation(self, *args, **kwargs):
...
+        if cache_implementation.upper() not in CacheImplementation.__members__:
+            raise ValueError(
+                f"Unrecognized cache implementation - you passed {cache_implementation}. Supported cache implementations are"
+                f" {CacheImplementation.__members__}"
+            )
+
+        if not self._supports_cache_class:
+            raise ValueError(
+                "This model do not currently support switching between cache implementations. Please raise a feature request on GitHub for adding"
+                " the request to support this model: https://github.com/huggingface/transformers"
```
Comment (tomaarsen) [inline src/transformers/generation/utils.py:384]:
```suggestion
                "You cannot currently update the cache implementation for this model. Please raise a feature request on GitHub for adding"
                " improved cache support this model: https://github.com/huggingface/transformers"
```
Maybe something like that?

File: src/transformers/generation/utils.py [hunk 8]
```diff
@@ -351,6 +359,43 @@ def prepare_inputs_for_generation(self, *args, **kwargs):
...
+            )
+
+        if cache_implementation == CacheImplementation.SINK:
+            if "window_length" not in kwargs and "num_sink_tokens" not in kwargs:
+                raise ValueError(
+                    "You asked to switch to the Sink cache implementation, but you did not pass `window_length` and `num_sink_tokens` to "
```
Comment (tomaarsen) [inline src/transformers/generation/utils.py:390]:
```suggestion
                    "You requested to use the Sink cache implementation, but you did not pass `window_length` and `num_sink_tokens` to "
```

**Review by tomaarsen — COMMENTED** [review]
Some more small nits regarding the phrasing now that we're using `set_...`

File: src/transformers/generation/utils.py [hunk 9]
```diff
@@ -351,6 +359,43 @@ def prepare_inputs_for_generation(self, *args, **kwargs):
             "A model class needs to define a `prepare_inputs_for_generation` method in order to use `.generate()`."
         )
 
+    def set_cache_implementation(self, cache_implementation: Union[CacheImplementation, str], **kwargs):
```
Comment (gante) [inline src/transformers/generation/utils.py:362]: I'd rather have this function as `set_cache`, with `past_key_values` as an optional parameter. But this is a tougher discussion: I've opened the discussion on our internal slack [here](https://huggingface.slack.com/archives/C060NPVEW57/p1708006459508009) 🤗

**Review by gante — COMMENTED** [review]
Much in favor of this! 👍

File: src/transformers/generation/utils.py [hunk 10]
```diff
@@ -351,6 +359,43 @@ def prepare_inputs_for_generation(self, *args, **kwargs):
...
+
+        if cache_implementation == CacheImplementation.SINK:
+            if "window_length" not in kwargs and "num_sink_tokens" not in kwargs:
+                raise ValueError(
+                    "You asked to switch to the Sink cache implementation, but you did not pass `window_length` and `num_sink_tokens` to "
+                    "`set_cache_implementation`. Try again with passing these arguments to the method. (e.g. `model.set_cache_implementation('sink', window_length=window_length=508, num_sink_tokens=4)`"
+                )
+            else:
+                self.generation_config.sink_window_length = kwargs.get("window_length")
+                self.generation_config.num_sink_tokens = kwargs.get("num_sink_tokens")
```
Comment (amyeroberts) [inline src/transformers/generation/utils.py:395]: No need for the else here - the error raise above is akin to an early return 

```suggestion
            self.generation_config.sink_window_length = kwargs.get("window_length")
            self.generation_config.num_sink_tokens = kwargs.get("num_sink_tokens")
```

Comment (amyeroberts) [inline src/transformers/generation/utils.py:395]: Just for my own understanding, if the currently set cache implementation is sink, with e.g. `window_length=window_length=508, num_sink_tokens=4`, should I be allowed to call `set_cache_implementation` with updated parameters i.e.

```
config.set_cache_implementation(CacheImplementation.SINK, window_length=1_016, num_sink_tokens=8)
```
?

**Review by amyeroberts — COMMENTED** [review]
Looks great! Thanks for adding

Agreed `set_` is preferable to `switch_`.

Only request is for tests!

File: src/transformers/generation/utils.py (hunk 5 above)
Comment (younesbelkada) [reply src/transformers/generation/utils.py]: Makes sense! Done !

File: src/transformers/generation/utils.py (hunk 10 above)
Comment (younesbelkada) [reply src/transformers/generation/utils.py]: yes I think so !
Per my understanding the way the API is designed currently for sink cache is to pass a `SinkCache()` object to `generate()` through `past_key_values`. If `past_key_values` is already passed in `generate()`, the way I designed things in the PR, we only use the cache from `set_cache_implementation` in case `past_key_values` is not passed to `generate`.

TLDR: If one calls `model.set_cache_implementation(CacheImplementation.SINK)`, they shouldn't call `model.generate(xxx, past_key_values=SinkCache())` --> maybe we should raise a warning saying that they already called `model.set_cache_impl()`, wdyt?

File: src/transformers/generation/utils.py [hunk 11]
```diff
@@ -92,6 +93,13 @@
 if is_accelerate_available():
     from accelerate.hooks import AlignDevicesHook, add_hook_to_module
 
+
+class CacheImplementation(str, Enum):
+    DYNAMIC = "dynamic"
+    STATIC = "static"
+    SINK = "sink"
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:100]: this prevents anyone from adding / using a custom implementation why not just use strings?

File: src/transformers/generation/utils.py [hunk 12]
```diff
@@ -1194,6 +1238,24 @@ def _validate_generated_length(self, generation_config, input_ids_length, has_de
...
+    def _sanitize_kwargs(self, generation_config: GenerationConfig, kwargs: dict):
+        """
+        Sanitize the kwargs for generation
+        """
+        if getattr(generation_config, "cache_implementation", "") == CacheImplementation.SINK:
+            if "past_key_values" not in kwargs:
+                kwargs["past_key_values"] = SinkCache(
+                    window_length=generation_config.sink_window_length,
+                    num_sink_tokens=generation_config.num_sink_tokens,
+                )
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:1251]: why not pass the class and do the same init scheme with generation config passed and you take these from the generation config to allow custom classes?

**Review by ArthurZucker — COMMENTED** [review]
Down for that, let's not add enums where we don't need, let's make it easy to use custom CacheClasses and init them with a generation config no?

File: src/transformers/generation/utils.py (hunk 11 above)
Comment (younesbelkada) [reply src/transformers/generation/utils.py:100]: I think in general it is a good practice to use enums to avoid silent behaviours / errors. 
E.g. if one passes 
```python
model.generation_config.cache_implementation = "Static"
```
The code will silently work out of the box as it will use dynamic cache, and can potentitally lead to silent errors. 

When using enums, 

```python
model.set_cache_implementation("Static")
```
Will also work out of the box but this time will correctly use the static cache implementation and not dynamic cache as opposed to the snippet above.

We can also use a mapping with hardcoded strings but I found it clearer to have enums 

cc @amyeroberts @gante @tomaarsen wdyt?

Comment (ArthurZucker) [reply src/transformers/generation/utils.py:100]: good practice does not mean it should always be used.
Here it's: 
- useless: we need a mapping with keys from `"string_cls":cls`
- cumbersome: anywhere you compare the generation_config.cache_cls (a string) you need to use the enum. Why? 
- not restrictive, and let me be clear here I do NOT want users to pass anything and expect it to work. If the class is not in the Mapping error out and we are done with it. 

TLDR; why would we use then when we don't need it and it only adds additional calls to `CacheImplementation.DYNAMIC` vs `"dynamic"` is something I don't understand.

Comment (amyeroberts) [reply src/transformers/generation/utils.py:100]: @ArthurZucker I don't understand your objections here. For one, best practices does mean it should be done wherever possible.

I don't believe using enums is useless or cumbersome (aside from these not being well defined)? As @younesbelkada highlights, they provide better guarantees of selection of valid types. The explicit enum means the user can pass a string, but the code uses stricter checks that string matching, which are error prone. 

> not restrictive, and let me be clear here I do NOT want users to pass anything and expect it to work. If the class is not in the Mapping error out and we are done with it. We should raise and error

Enums are restrictive? In fact, they're far more restrictive than doing string checks and have stronger guarantee's than checking a mapping which is mutable.

Comment (ArthurZucker) [reply src/transformers/generation/utils.py:100]: But that is exactly what we want: the mapping should be mutable, because we want it to be easily adapted for a custom code on the hub. 
Down for stricter checks, but in this specific case IMO it is cumbersome and useless as we don't have the notion of "safety" and we basically use mapping for model_type, config_type etc etc.

Comment (ArthurZucker) [reply src/transformers/generation/utils.py:100]: Anyway it's not that important

Comment (amyeroberts) [reply src/transformers/generation/utils.py:100]: Ah, OK, I understand better now. I'm not sure how we can both let users modify a dict and have guarantees about not working if users can pass anything. If I've understood correctly, is the check you're wanting on the membership within the dictionary rather than e.g. how it's handled within the setting logic? In this case, I agree dictionaries are probably the simplest solution.

Comment (ArthurZucker) [reply src/transformers/generation/utils.py:100]: Yes exactly 😉
